lxd/placement/filtering.go

```diff
@@ -0,0 +1,118 @@
+package placement
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/http"
+	"slices"
+	"strings"
+
+	"github.com/canonical/lxd/lxd/db"
+	"github.com/canonical/lxd/lxd/db/cluster"
+	"github.com/canonical/lxd/lxd/db/query"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/entity"
+)
+
+// ApplyRuleset checks if the given configuration contains a placement ruleset. If not, it returns the candidates
+// unaltered. If so, it filters the candidates via the found ruleset.
+func ApplyRuleset(ctx context.Context, tx *sql.Tx, candidates []db.NodeInfo, expandedConfig map[string]string, runningInstanceID *int, ruleset cluster.PlacementRuleset) ([]db.NodeInfo, error) {
+	rules := ruleset.SortedRules()
+
+	entityIDsToMemberIDs := func(entityType entity.Type, entityIDs []int) ([]int, error) {
+		args := make([]any, 0, len(entityIDs))
+		for _, id := range entityIDs {
+			args = append(args, id)
+		}
+
+		switch entityType {
+		case entity.TypeInstance:
+			return query.SelectIntegers(ctx, tx, `SELECT DISTINCT node_id FROM instances WHERE id IN `+query.Params(len(entityIDs)), args...)
+		case entity.TypeClusterGroup:
+			return query.SelectIntegers(ctx, tx, `SELECT DISTINCT node_id FROM nodes_cluster_groups WHERE group_id IN `+query.Params(len(entityIDs)), args...)
+		case entity.TypeClusterMember:
+			return entityIDs, nil
+		default:
+			return nil, fmt.Errorf("Invalid placement rule selector entity type %q", entityType)
+		}
+	}
+
+	isSelfMatch := func(selector cluster.Selector) bool {
+		if entity.Type(selector.EntityType) != entity.TypeInstance {
+			return false
+		}
+
+		for _, m := range selector.Matchers {
+			_, key, ok := strings.Cut(m.Property, "config.")
```

Comment:
`strings.CutPrefix`?

Reply:
copy/pasto again from a time before I knew that CutPrefix existed ;)